stdlib/camlinternalStorage.ml

```diff
@@ -0,0 +1,214 @@
+module Obj_opt : sig
+  type t
+  val none : t
+  val some : 'a -> t
+  val is_some : t -> bool
+
+  (** [unsafe_get obj] may only be called safely
+      if [is_some] is true.
+
+      [unsafe_get (some v)] is equivalent to
+      [Obj.obj (Obj.repr v)]. *)
+  val unsafe_get : t -> 'a
+end = struct
+  type t = Obj.t
+  let none = Obj.repr (ref 0)
+  let some v = Obj.repr v
+  let is_some obj = (obj != none)
+  let unsafe_get obj = Obj.obj obj
+end
+
+type state = Obj_opt.t array
+
+module type State_operations = sig
+  val get : unit -> state
+  val compare_and_set : state -> state -> bool
+end
+
+module type Key = sig
+    type 'a t
+
+    val make : ?split_from_parent:('a -> 'a) -> (unit -> 'a) -> 'a t
+
+    val get : 'a t -> 'a
+
+    val set : 'a t -> 'a -> unit
+
+    type key_value = KV : 'a t * 'a -> key_value
+    val get_initial_keys : unit -> key_value list
+    val set_initial_keys : key_value list -> unit
+end
+
+module Make (State : State_operations)
+  : Key with type 'a t = int * (unit -> 'a)
+= struct
+  type 'a t = int * (unit -> 'a)
+
+  (* Manipulating existing keys. *)
+
+  (* If necessary, grow the current domain's local state array such that [idx]
+   * is a valid index in the array. *)
+  let rec maybe_grow idx =
+    let st = State.get () in
+    let sz = Array.length st in
+    if idx < sz then st
+    else begin
+      let rec compute_new_size s =
+        if idx < s then s else compute_new_size (2 * s)
+      in
+      let new_sz = compute_new_size (max sz 8) in
```

Comment:
```suggestion
        if idx < s then s else compute_new_size (2 * s + 1)
      in
      let new_sz = compute_new_size (max sz 7) in
```

To keep the size as power of two minus one such that the whole block is a power of two and less likely to suffer from false sharing.

Are there other places that need this adjustment?